File: baseline/tech-writer.py
from pathlib import Path
import json
import re
from openai import OpenAI
import inspect
import typing
import sys

# Import from common/utils.py
from common.utils import (
    read_prompt_file,
    save_results,
    create_metadata,
    REACT_SYSTEM_PROMPT,
    CustomEncoder,
    validate_github_url,
    clone_repo,
    get_command_line_args,
    OPENAI_API_KEY,
    GEMINI_API_KEY,
    GEMINI_MODELS
)

from common.tools import TOOLS
from common.logging import logger, configure_logging
class TechWriterReActAgent:
    def __init__(self, model_name="gpt-4.1-mini", base_url=None):
        """Initialise the agent with the specified model."""
        self.model_name = model_name
        self.memory = []
        self.final_answer = None
        self.system_prompt = None  # To be defined by subclasses
        
        # Determine which API to use based on the model name
        if model_name in GEMINI_MODELS:
            if not GEMINI_API_KEY:
                raise ValueError("GEMINI_API_KEY environment variable is not set but a Gemini model was specified.")
            self.client = OpenAI(
                api_key=GEMINI_API_KEY,
                base_url=base_url or "https://generativelanguage.googleapis.com/v1beta/openai/"
            )
        else:
            if not OPENAI_API_KEY:
                raise ValueError("OPENAI_API_KEY environment variable is not set but an OpenAI model was specified.")
            self.client = OpenAI(api_key=OPENAI_API_KEY, base_url=base_url)

        """Initialise the ReAct agent with the specified model."""
        self.system_prompt = REACT_SYSTEM_PROMPT
   
    def create_openai_tool_definitions(self, tools_dict):
        """
        Create tool definitions from a dictionary of Python functions.
        
        Args:
            tools_dict: Dictionary mapping tool names to Python functions
            
        Returns:
            List of tool definitions formatted for the OpenAI API
        """
        tools = []
        
        for name, func in tools_dict.items():
            # Extract function signature
            sig = inspect.signature(func)
            
            # Get docstring and parse it
            docstring = inspect.getdoc(func) or ""
            description = docstring.split("\n\n")[0] if docstring else ""
            
            # Build parameters
            parameters = {
                "type": "object",
                "properties": {},
                "required": []
            }
            
            for param_name, param in sig.parameters.items():
                if param_name == "self":
                    continue
                
                param_type = param.annotation
                if param_type is inspect.Parameter.empty:
                    param_type = str
                
                # Get origin and args for generic types
                origin = typing.get_origin(param_type)
                args = typing.get_args(param_type)
                
                # Convert Python types to JSON Schema types
                if param_type == str:
                    json_type = "string"
                elif param_type == int:
                    json_type = "integer"
                elif param_type == float or param_type == "number":
                    json_type = "number"
                elif param_type == bool:
                    json_type = "boolean"
                elif origin is list or param_type == list:
                    json_type = "array"
                elif origin is dict or param_type == dict:
                    json_type = "object"
                else:
                    # For complex types, default to string
                    json_type = "string"
                
                # Extract parameter description from docstring
                param_desc = ""
                if docstring:
                    # Look for parameter in docstring (format: param_name: description)
                    param_pattern = rf"{param_name}:\s*(.*?)(?:\n\s*\w+:|$)"
                    param_match = re.search(param_pattern, docstring, re.DOTALL)
                    if param_match:
                        param_desc = param_match.group(1).strip()
                
                # Add parameter to schema
                parameters["properties"][param_name] = {
                    "type": json_type,
                    "description": param_desc
                }
                
                # Mark required parameters
                if param.default is inspect.Parameter.empty:
                    parameters["required"].append(param_name)
            
            # Create tool definition
            tool_def = {
                "type": "function",
                "function": {
                    "name": name,
                    "description": description,
                    "parameters": parameters
                }
            }
            
            tools.append(tool_def)
        
        return tools
    
    def initialise_memory(self, prompt, directory):
        """Initialise the agent's memory with the prompt and directory."""
        if not self.system_prompt:
            raise ValueError("System prompt must be defined by subclasses")
            
        self.memory = [{"role": "system", "content": self.system_prompt}]
        self.memory.append({"role": "user", "content": f"Base directory: {directory}\n\n{prompt}"})
        self.final_answer = None
    
    def call_llm(self):
        """
        Call the LLM with the current memory and tools.
        
        Uses the OpenAI client with appropriate base_url for all models.
        """
        try:
            response = self.client.chat.completions.create(
                model=self.model_name,
                messages=self.memory,
                tools=self.create_openai_tool_definitions(TOOLS),
                temperature=0
            )
            return response.choices[0].message
        except Exception as e:
            error_msg = f"Error calling API: {str(e)}"
            logger.error(error_msg)
            raise ValueError(error_msg)
    
    def check_llm_result(self, assistant_message):
        """
        Check if the LLM result is a final answer or a tool call.
        
        Args:
            assistant_message: The message from the assistant
            
        Returns:
            tuple: (result_type, result_data)
                result_type: "final_answer" or "tool_calls"
                result_data: The final answer string or list of tool calls
        """
        self.memory.append(assistant_message)
        
        if assistant_message.tool_calls:
            return "tool_calls", assistant_message.tool_calls
        else:
            return "final_answer", assistant_message.content
    
    def execute_tool(self, tool_call):
        """
        Execute a tool call and return the result.
        
        Args:
            tool_call: The tool call object from the LLM
            
        Returns:
            str: The result of the tool execution
        """
        tool_name = tool_call.function.name
        
        if tool_name not in TOOLS:
            return f"Error: Unknown tool {tool_name}"
        
        try:
            # Parse the arguments
            args = json.loads(tool_call.function.arguments)
            
            # Call the tool function
            result = TOOLS[tool_name](**args)
            
            # Convert result to JSON string
            return json.dumps(result, cls=CustomEncoder, indent=2)
        except json.JSONDecodeError as e:
            return f"Error: Invalid JSON in tool arguments: {str(e)}"
        except TypeError as e:
            return f"Error: Invalid argument types: {str(e)}"
        except ValueError as e:
            return f"Error: Invalid argument values: {str(e)}"
        except Exception as e:
            return f"Error executing tool {tool_name}: {str(e)}"
    

    def run(self, prompt, directory):
        """Run the agent to analyse a codebase using the ReAct pattern."""
        self.initialise_memory(prompt, directory)
        max_steps = 15
        
        for step in range(max_steps):
            logger.info(f"\n--- Step {step + 1} ---")
            logger.debug(f"Current memory size: {len(self.memory)} messages")
            
            # Call the LLM
            try:
                assistant_message = self.call_llm()
                
                # Check the result
                result_type, result_data = self.check_llm_result(assistant_message)
                
                if result_type == "final_answer":
                    logger.info("Received final answer from LLM")
                    self.final_answer = result_data
                    break
                elif result_type == "tool_calls":
                    logger.info(f"Processing {len(result_data)} tool calls")
                    # Execute each tool call
                    for tool_call in result_data:
                        logger.debug(f"Executing tool: {tool_call.function.name} with args: {tool_call.function.arguments}")
                        # Execute the tool
                        observation = self.execute_tool(tool_call)
                        logger.debug(f"Tool result length: {len(observation) if observation else 0} chars")
                        
                        # Add the observation to memory
                        self.memory.append({
                            "role": "tool",
                            "tool_call_id": tool_call.id,
                            "name": tool_call.function.name,
                            "content": observation
                        })
            except Exception as e:
                logger.error(f"Unexpected error in step {step + 1}: {e}", exc_info=True)
                self.final_answer = f"Error running code analysis: {e}"
                break
            
            logger.info(f"Memory length: {len(self.memory)} messages")
            logger.debug(f"Current memory content size: {sum(len(str(msg)) for msg in self.memory)} chars")
        
        if self.final_answer is None:
            logger.warning(f"Failed to complete analysis within {max_steps} steps")
            self.final_answer = "Failed to complete the analysis within the step limit."
        
        return self.final_answer


def analyse_codebase(directory_path: str, prompt_file_path: str, model_name: str, base_url: str = None, repo_url: str = None) -> tuple[str, str, str]:
    """
    Analyse a codebase using the specified agent type with a prompt from an external file.
    
    Args:
        directory_path: Path to directory containing codebase
        prompt_file_path: Path to file containing analysis prompt
        model_name: Name of model to use for analysis
        base_url: Base URL for API (optional)
        repo_url: GitHub repository URL if cloned from GitHub (optional)
        
    Returns:
        tuple: (analysis_result, repo_name, repo_url)
    """
    # Read the prompt from file
    prompt = read_prompt_file(prompt_file_path)
    
    agent = TechWriterReActAgent(model_name, base_url)
    
    # Run the analysis
    analysis_result = agent.run(prompt, directory_path)
    
    # Get repository name for output file
    repo_name = Path(directory_path).name
    
    return analysis_result, repo_name, repo_url or ""


def main():
    try:
        configure_logging()
        args = get_command_line_args()
        
        # Handle repo cloning if specified
        repo_url = ""
        if args.repo:
            if not validate_github_url(args.repo):
                raise ValueError("Invalid GitHub repository URL format")
            try:
                directory_path = str(clone_repo(args.repo, args.cache_dir))
                repo_url = args.repo
            except Exception as e:
                logger.error(f"Failed to clone repository: {str(e)}")
                sys.exit(1)
        else:
            directory_path = args.directory
            
        # Validate directory exists
        if not Path(directory_path).exists():
            raise FileNotFoundError(f"Directory not found: {directory_path}")
            
        analysis_result, repo_name, _ = analyse_codebase(directory_path, args.prompt_file, args.model, args.base_url, repo_url)
        
        # Check if the result is an error message or a step limit failure
        if isinstance(analysis_result, str) and \
           (analysis_result.startswith("Error running code analysis:") or \
            analysis_result == "Failed to complete the analysis within the step limit."):
            logger.error(analysis_result)
            sys.exit(1)
        
        # Save the results
        output_file = save_results(analysis_result, args.model, repo_name, args.output_dir, args.extension)
        logger.info(f"Analysis complete. Results saved to: {output_file}")
        
        # Always create metadata (with optional evaluation)
        create_metadata(output_file, args.model, repo_url, repo_name, analysis_result, args.eval_prompt)
        
    except Exception as e:
        logger.error(f"Error: {str(e)}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: noframework/php/tech-writer.php
#!/usr/bin/env php
<?php
require_once __DIR__ . '/vendor/autoload.php';

use GuzzleHttp\Client;
use Monolog\Logger;
use Monolog\Handler\StreamHandler;
use Symfony\Component\Finder\Finder;

// Constants
const MAX_STEPS = 50;
const TEMPERATURE = 0;
const TEXT_MIME_TYPES = ['application/json', 'application/xml', 'application/javascript'];
const REACT_SYSTEM_PROMPT = 'You are a technical documentation assistant that analyses codebases and generates comprehensive documentation.

When given a directory path and a specific analysis request, you will:
1. Explore the codebase structure to understand its organization
2. Read relevant files to comprehend the implementation details
3. Generate detailed technical documentation based on your analysis

You have access to tools that help you explore and understand codebases:
- find_all_matching_files: Find files matching patterns in directories
- read_file: Read the contents of specific files

Important guidelines:
- Always start by exploring the directory structure to understand the codebase layout
- Read files strategically based on the documentation needs
- Pay attention to configuration files, main entry points, and key modules
- Generate clear, well-structured documentation that would help developers understand the codebase

Use the following format:

Thought: I need to [describe what you need to do next]
Action: [tool_name]
Action Input: {"param1": "value1", "param2": "value2"}
Observation: [tool output will be provided here]
... (repeat Thought/Action/Action Input/Observation as needed)
Thought: I now have enough information to generate the documentation
Final Answer: [Your complete technical documentation]

Begin your analysis now.';

class TechWriterAgent {
    private $logger;
    private $client;
    private $modelId;
    private $memory = [];
    private $finalAnswer = null;
    private $baseUrl;
    private $apiKey;

    public function __construct($modelName, $baseUrl = null) {
        // Set up logging
        $logDir = __DIR__ . '/logs';
        if (!is_dir($logDir)) {
            mkdir($logDir, 0755, true);
        }
        $this->logger = new Logger('tech-writer');
        $this->logger->pushHandler(new StreamHandler($logDir . '/tech-writer-' . date('Ymd-His') . '.log', Logger::DEBUG));
        
        // Parse model name
        $parts = explode('/', $modelName, 2);
        if (count($parts) !== 2) {
            throw new Exception("Invalid model name format");
        }
        
        $vendor = $parts[0];
        $this->modelId = $parts[1];
        
        // Set up API client based on vendor
        switch ($vendor) {
            case 'google':
                $this->apiKey = getenv('GEMINI_API_KEY');
                if (!$this->apiKey) {
                    throw new Exception("GEMINI_API_KEY environment variable is not set");
                }
                $this->baseUrl = $baseUrl ?: 'https://generativelanguage.googleapis.com/v1beta/openai/';
                break;
            
            case 'openai':
                $this->apiKey = getenv('OPENAI_API_KEY');
                if (!$this->apiKey) {
                    throw new Exception("OPENAI_API_KEY environment variable is not set");
                }
                $this->baseUrl = $baseUrl ?: 'https://api.openai.com/v1/';
                break;
            
            default:
                throw new Exception("Unknown model vendor: $vendor");
        }
        
        $this->client = new Client([
            'base_uri' => $this->baseUrl,
            'headers' => [
                'Authorization' => 'Bearer ' . $this->apiKey,
                'Content-Type' => 'application/json'
            ]
        ]);
    }

    public function run($prompt, $directory) {
        $this->logger->info("Starting ReAct agent with model: {$this->modelId}");
        
        // Initialize conversation
        $this->memory = [
            ['role' => 'system', 'content' => REACT_SYSTEM_PROMPT],
            ['role' => 'user', 'content' => "Base directory for analysis: $directory\n\n$prompt"]
        ];
        
        // ReAct loop
        for ($step = 0; $step < MAX_STEPS; $step++) {
            $this->logger->info("Step " . ($step + 1) . "/" . MAX_STEPS);
            
            try {
                // Get LLM response
                $response = $this->callLLM();
                $this->logger->debug("LLM Response: $response");
                
                // Add assistant response to memory
                $this->memory[] = ['role' => 'assistant', 'content' => $response];
                
                // Parse response
                $parsed = $this->parseResponse($response);
                
                if ($parsed['type'] === 'final') {
                    $this->finalAnswer = $parsed['answer'];
                    $this->logger->info("Final answer received");
                    break;
                } elseif ($parsed['type'] === 'action') {
                    // Execute tool
                    $observation = $this->executeTool($parsed['action'], $parsed['input']);
                    
                    // Add observation to memory
                    $this->memory[] = ['role' => 'user', 'content' => "Observation: $observation"];
                    
                    $this->logger->debug("Tool result: " . strlen($observation) . " chars");
                }
            } catch (Exception $e) {
                $this->logger->error("Error in step " . ($step + 1) . ": " . $e->getMessage());
                throw $e;
            }
        }
        
        if ($this->finalAnswer === null) {
            throw new Exception("Failed to complete analysis within " . MAX_STEPS . " steps");
        }
        
        return $this->finalAnswer;
    }

    private function callLLM() {
        $response = $this->client->post('chat/completions', [
            'json' => [
                'model' => $this->modelId,
                'messages' => $this->memory,
                'temperature' => TEMPERATURE
            ]
        ]);
        
        $data = json_decode($response->getBody(), true);
        return $data['choices'][0]['message']['content'] ?? '';
    }

    private function parseResponse($response) {
        // Check for Final Answer
        if (strpos($response, 'Final Answer:') !== false) {
            $pos = strpos($response, 'Final Answer:');
            $finalAnswer = trim(substr($response, $pos + 13));
            return ['type' => 'final', 'answer' => $finalAnswer];
        }
        
        // Extract Action and Action Input
        $lines = explode("\n", $response);
        $action = null;
        $collectingInput = false;
        $inputLines = [];
        
        foreach ($lines as $line) {
            if (strpos($line, 'Action:') === 0) {
                $action = trim(substr($line, 7));
            } elseif (strpos($line, 'Action Input:') === 0) {
                $collectingInput = true;
                $rest = trim(substr($line, 13));
                if (!empty($rest)) {
                    $inputLines[] = $rest;
                }
            } elseif ($collectingInput) {
                // Stop collecting if we hit another section marker
                if (strpos($line, 'Thought:') === 0 || 
                    strpos($line, 'Action:') === 0 || 
                    strpos($line, 'Observation:') === 0 || 
                    strpos($line, 'Final Answer:') === 0) {
                    break;
                }
                $inputLines[] = $line;
            }
        }
        
        if ($action && !empty($inputLines)) {
            $input = trim(implode("\n", $inputLines));
            return ['type' => 'action', 'action' => $action, 'input' => $input];
        }
        
        return ['type' => 'unknown'];
    }

    private function executeTool($toolName, $actionInput) {
        $this->logger->debug("Executing tool: $toolName with input: $actionInput");
        
        $input = json_decode($actionInput, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            return json_encode(['error' => 'Error parsing input: ' . json_last_error_msg()]);
        }
        
        switch ($toolName) {
            case 'find_all_matching_files':
                $directory = $input['directory'] ?? '';
                $pattern = $input['pattern'] ?? '*';
                return json_encode($this->findAllMatchingFiles($directory, $pattern));
                
            case 'read_file':
                $filePath = $input['file_path'] ?? '';
                return json_encode($this->readFile($filePath));
                
            default:
                return json_encode(['error' => "Unknown tool: $toolName"]);
        }
    }

    private function findAllMatchingFiles($directory, $pattern) {
        $this->logger->info("Tool invoked: find_all_matching_files(directory='$directory', pattern='$pattern')");
        
        if (!is_dir($directory)) {
            $this->logger->error("Directory not found: $directory");
            return [];
        }
        
        $finder = new Finder();
        $finder->files()->in($directory);
        
        // Include dot files (hidden files) - let gitignore handle exclusions
        $finder->ignoreDotFiles(false);
        
        // Always exclude .git directory
        $finder->exclude('.git');
        
        // Handle pattern
        if ($pattern === '*' || $pattern === '*.*') {
            // Match all files
        } elseif (strpos($pattern, '*.') === 0) {
            // Extension pattern
            $ext = substr($pattern, 2);
            $finder->name("*.$ext");
        } else {
            // Exact name
            $finder->name($pattern);
        }
        
        // Respect .gitignore
        if (file_exists($directory . '/.gitignore')) {
            $gitignore = file_get_contents($directory . '/.gitignore');
            $patterns = array_filter(explode("\n", $gitignore), function($line) {
                return !empty($line) && $line[0] !== '#';
            });
            foreach ($patterns as $ignorePattern) {
                $finder->notPath($ignorePattern);
            }
        }
        
        $files = [];
        foreach ($finder as $file) {
            $files[] = $file->getRealPath();
        }
        
        $this->logger->info("Found " . count($files) . " matching files");
        return $files;
    }

    private function readFile($filePath) {
        $this->logger->info("Tool invoked: read_file(file_path='$filePath')");
        
        if (!file_exists($filePath)) {
            return ['error' => "File not found: $filePath"];
        }
        
        // Check if binary
        $finfo = finfo_open(FILEINFO_MIME_TYPE);
        $mimeType = finfo_file($finfo, $filePath);
        finfo_close($finfo);
        
        if (strpos($mimeType, 'text/') !== 0 && !in_array($mimeType, TEXT_MIME_TYPES)) {
            $this->logger->debug("File detected as binary: $filePath");
            return ['error' => "Cannot read binary file: $filePath"];
        }
        
        $content = file_get_contents($filePath);
        if ($content === false) {
            return ['error' => "Failed to read file: $filePath"];
        }
        
        $charCount = strlen($content);
        $this->logger->info("Successfully read file: $filePath ($charCount chars)");
        
        return [
            'file' => $filePath,
            'content' => $content
        ];
    }
}

// Helper functions
function sanitizeFilename($name) {
    return str_replace(['/', '\\', ':', '*', '?', '"', '<', '>', '|'], '-', $name);
}

function saveResults($content, $repoName, $model, $outputDir, $extension, $fileName = null) {
    global $logger;
    
    if (!is_dir($outputDir)) {
        mkdir($outputDir, 0755, true);
    }
    
    if ($fileName) {
        $outputPath = $outputDir . '/' . $fileName;
    } else {
        $timestamp = date('Ymd-His');
        $parts = explode('/', $model, 2);
        $vendor = $parts[0];
        $modelId = sanitizeFilename($parts[1]);
        
        $outputPath = $outputDir . "/$timestamp-$repoName-$vendor-$modelId$extension";
    }
    
    file_put_contents($outputPath, $content);
    $logger->info("Results saved to: $outputPath");
    
    return $outputPath;
}

function createMetadata($outputFile, $model, $repoUrl, $repoName) {
    global $logger;
    
    $metadataFile = preg_replace('/\.[^.]+$/', '.metadata.json', $outputFile);
    
    $metadata = [
        'model' => $model,
        'github_url' => $repoUrl,
        'repo_name' => $repoName,
        'timestamp' => date('c')
    ];
    
    file_put_contents($metadataFile, json_encode($metadata, JSON_PRETTY_PRINT));
    $logger->info("Metadata saved to: $metadataFile");
}

function cloneOrUpdateRepo($repoUrl, $cacheDir) {
    global $logger;
    
    $parts = explode('/', rtrim($repoUrl, '/'));
    $repoName = str_replace('.git', '', end($parts));
    $owner = $parts[count($parts) - 2];
    
    $cacheDir = str_replace('~', $_SERVER['HOME'], $cacheDir);
    $cachePath = "$cacheDir/$owner/$repoName";
    
    if (!is_dir(dirname($cachePath))) {
        mkdir(dirname($cachePath), 0755, true);
    }
    
    if (is_dir("$cachePath/.git")) {
        $logger->info("Updating existing repository: $cachePath");
        exec("cd '$cachePath' && git pull --quiet");
    } else {
        $logger->info("Cloning repository: $repoUrl");
        exec("git clone --quiet '$repoUrl' '$cachePath'");
    }
    
    return $cachePath;
}

// Parse command line arguments
$options = getopt('h', [
    'repo:',
    'prompt:',
    'cache-dir:',
    'output-dir:',
    'extension:',
    'file-name:',
    'model:',
    'base-url:',
    'help'
]);

// Show help
if (isset($options['h']) || isset($options['help'])) {
    echo "Usage: $argv[0] [directory] [options]\n\n";
    echo "Analyse a codebase using an LLM agent.\n\n";
    echo "Positional arguments:\n";
    echo "  directory             Directory containing the codebase to analyse\n\n";
    echo "Options:\n";
    echo "  --repo REPO           GitHub repository URL to clone (e.g. https://github.com/owner/repo)\n";
    echo "  --prompt FILE         Path to a file containing the analysis prompt (required)\n";
    echo "  --cache-dir DIR       Directory to cache cloned repositories (default: ~/.cache/github)\n";
    echo "  --output-dir DIR      Directory to save results to (default: output)\n";
    echo "  --extension EXT       File extension for output files (default: .md)\n";
    echo "  --file-name FILE      Specific file name for output (overrides --extension)\n";
    echo "  --model MODEL         Model to use (format: vendor/model, default: openai/gpt-4o-mini)\n";
    echo "  --base-url URL        Base URL for the API (automatically set based on model if not provided)\n";
    echo "  -h, --help            Show this help message and exit\n\n";
    echo "Dependencies:\n";
    echo "  This script requires environment variables:\n";
    echo "  - OPENAI_API_KEY for OpenAI models\n";
    echo "  - GEMINI_API_KEY for Google models\n";
    exit(0);
}

// Get positional directory argument
$positionalArgs = array_filter($argv, function($arg) {
    return $arg[0] !== '-';
});
array_shift($positionalArgs); // Remove script name
$directory = array_shift($positionalArgs);

// Set defaults
$promptFile = $options['prompt'] ?? null;
$cacheDir = $options['cache-dir'] ?? '~/.cache/github';
$outputDir = $options['output-dir'] ?? 'output';
$extension = $options['extension'] ?? '.md';
$fileName = $options['file-name'] ?? null;
$model = $options['model'] ?? 'openai/gpt-4o-mini';
$baseUrl = $options['base-url'] ?? null;

// Validate arguments
if (!$promptFile) {
    echo "Error: --prompt is required\n";
    exit(1);
}

if (!file_exists($promptFile)) {
    echo "Error: Prompt file not found: $promptFile\n";
    exit(1);
}

// Read prompt
$prompt = file_get_contents($promptFile);

// Set up logging
$logger = new Logger('tech-writer');
$logDir = __DIR__ . '/logs';
if (!is_dir($logDir)) {
    mkdir($logDir, 0755, true);
}
$logger->pushHandler(new StreamHandler($logDir . '/tech-writer-' . date('Ymd-His') . '.log', Logger::DEBUG));
$logger->pushHandler(new StreamHandler('php://stderr', Logger::INFO));

try {
    // Handle repository or directory
    if (isset($options['repo'])) {
        $repoUrl = $options['repo'];
        $directory = cloneOrUpdateRepo($repoUrl, $cacheDir);
        $repoName = basename($repoUrl, '.git');
    } else {
        if (!$directory) {
            $directory = '.';
        }
        if (!is_dir($directory)) {
            echo "Error: Directory not found: $directory\n";
            exit(1);
        }
        $directory = realpath($directory);
        $repoName = basename($directory);
        $repoUrl = '';
    }
    
    // Run the agent
    $agent = new TechWriterAgent($model, $baseUrl);
    $analysisResult = $agent->run($prompt, $directory);
    
    // Save results
    $outputFile = saveResults($analysisResult, $repoName, $model, $outputDir, $extension, $fileName);
    
    // Create metadata
    createMetadata($outputFile, $model, $repoUrl, $repoName);
    
} catch (Exception $e) {
    $logger->error("Analysis failed: " . $e->getMessage());
    exit(1);
}